woocommerce/woocommerce
# Ensure Product Collection Block's `queryId` Uniqueness

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

One of the caveats to block duplication is that _all_ attributes are copied. This has caused some trouble because the `queryId` are expected to be unique. This PR makes some changes to the block initialization in able to ensure that both attributes have a unique value.

- `queryId`: [This problem was also present in Gutenberg](https://github.com/WordPress/gutenberg/pull/58804) and I have applied the same fix to both this block and the Query Loop block. Since we only need the ID to remain stable across frontend page loads, we can just set `instanceId` each time. It will only _actually_ do anything if it changes so there's no problem with doing it this way.

Closes #43311



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new post.
2. Add a new "Product Collection" block and note the `queryId` in the Code Editor.
3. Add a new "Product Collection" block and confirm that the `queryId` are different in the Code Editor.
4. Duplicate the first "Product Collection" block and note that the `queryId` is different.
5. Refresh the page, the `queryId` should re-index and still ensure uniqueness.

**Note: The `queryId` tracks the number of instances that have been created on the page. Switching between the code editor and visual editor will not reset the counter and so after switching back you'll see the IDs have a weird offset. This is fixed on the next editor reload but doesn't actually hurt anything.**



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
